clippy_lints/src/operators/manual_midpoint.rs

```diff
@@ -0,0 +1,45 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::msrvs::{self, Msrv};
+use clippy_utils::sugg::Sugg;
+use clippy_utils::{is_floating_point_integer_literal, is_integer_literal};
+use rustc_ast::BinOpKind;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::LateContext;
+use rustc_middle::ty;
+
+use super::MANUAL_MIDPOINT;
+
+pub(super) fn check<'tcx>(
+    cx: &LateContext<'tcx>,
+    expr: &'tcx Expr<'_>,
+    op: BinOpKind,
+    left: &'tcx Expr<'_>,
+    right: &'tcx Expr<'_>,
+    msrv: &Msrv,
+) {
+    if msrv.meets(msrvs::UINT_FLOAT_MIDPOINT)
+        && op == BinOpKind::Div
+        && (is_integer_literal(right, 2) || is_floating_point_integer_literal(right, 2))
+        && let ExprKind::Binary(left_op, ll_expr, lr_expr) = left.kind
+        && left_op.node == BinOpKind::Add
+        && let left_ty = cx.typeck_results().expr_ty_adjusted(ll_expr)
+        && let right_ty = cx.typeck_results().expr_ty_adjusted(lr_expr)
+        && left_ty == right_ty
+        && matches!(left_ty.kind(), ty::Uint(_) | ty::Float(_))
+    {
+        let mut app = Applicability::MachineApplicable;
+        let left_sugg = Sugg::hir_with_applicability(cx, ll_expr, "..", &mut app);
+        let right_sugg = Sugg::hir_with_applicability(cx, lr_expr, "..", &mut app);
+        let sugg = format!("{left_ty}::midpoint({left_sugg}, {right_sugg})");
+        span_lint_and_sugg(
+            cx,
+            MANUAL_MIDPOINT,
+            expr.span,
+            "manual implementation of `midpoint`",
```

Comment:
```suggestion
            "manual implementation of `midpoint` which can overflow",
```

Reply:
I'm reluctant to add this on the lint message, because this particular use may well never overflow depending on the context. This is appropriate for the lint short description though, I'll add it there. What do you think?

Reply:
Any addition may overflow, and we do not suggest using `.checked_add()`. I'll let other weigh in, but I'm not comfortable saying "which may overflow" on `(a + b) / 2` if `a` and `b` are indices in a vector for example, they will never overflow as they will be nowhere near `usize::MAX/2` for any realistic vector.

Reply:
> `(a + b) / 2` if `a` and `b` are indices in a vector for example

`u8::MAX = 255`, `u16::MAX = 65536` are fairly realistic numbers to me.

> they will never overflow as they will be nowhere near `usize::MAX/2` for any realistic vector.

reminder of https://research.google/blog/extra-extra-read-all-about-it-nearly-all-binary-searches-and-mergesorts-are-broken/

Reply:
I was specifically talking about the `usize` case which is used to index a vector. And let's not confuse the issue: I'm not trying to argue that we should not lint, we will, only that I am not comfortable saying that a particular computation may overflow.

I couldn't find lints that warn about the potential risk when linting an expression if the risk is remote or non-existing for this particular expression, even though they still lint to make it clearer and safer should this code be changed later.